app/Services/AuthTokenService.php

```diff
@@ -0,0 +1,295 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Services;
+
+use App\Models\AuthToken;
+use App\Models\User;
+use CDash\Model\Project;
+use CDash\Model\UserProject;
+use InvalidArgumentException;
+
+class AuthTokenService
+{
+    /**
+     * Contract: we assume that $user_id has already been validated and blindly create a token
+     * for the user specified.  It is the responsibility of anyone who uses this function to
+     * ensure that the $user_id has been properly authenticated.
+     *
+     * @return array{raw_token: string, token: AuthToken}
+     */
+    public static function generateToken(int $user_id, int $project_id, string $scope, string $description): array
+    {
+        $token = bin2hex(random_bytes(16));
+        $params['hash'] = hash('sha512', $token);
+
+        $params['userid'] = $user_id;
+
+        $duration = config('cdash.token_duration');
+        $now = time();
+        $params['created'] = gmdate(FMT_DATETIME, $now);
+
+        if (!is_numeric($duration) || intval($duration) < 0) {
+            add_log("Invalid token_duration configuration {$duration}", 'AuthTokenService::generateToken', LOG_ERR);
```

Comment:
Can we use `\Log::error()` instead? `add_log()` is pre-Laravel cruft that we should probably deprecate and remove.